cloud/amazon/ec2_elb.py

```diff
@@ -176,64 +194,53 @@ def _enable_availailability_zone(self, lb):
-                    self.changed = True
+            instance_states = self._get_instance_health(lb)
+            instances_not_ok = copy.deepcopy(self.instance_ids) 
+            for instance_id,instance_state in instance_states.iteritems():
+                if instance_state.state == awaited_state:
+                    instances_not_ok.remove(instance_id)                    
+                    # Check the current state against the initial state, and only set
+                    # changed if they are different.
+                    if instance_state.state != initial_states[instance_id].state:
+                        self.changed = True
+                        self.elb_changed[lb.name].append(instance_id)
+
+            if len(instances_not_ok) == 0:
                 break
-            elif self._is_instance_state_pending(instance_state):
-                # If it's pending, we'll skip further checks andd continue waiting
-                pass
-            elif (awaited_state == 'InService'
-                  and instance_state.reason_code == "Instance"
-                  and time.time() >= wait_timeout):
-                # If the reason_code for the instance being out of service is
-                # "Instance" this indicates a failure state, e.g. the instance
-                # has failed a health check or the ELB does not have the
-                # instance's availabilty zone enabled. The exact reason why is
-                # described in InstantState.description.
-                msg = ("The instance %s could not be put in service on %s."
-                       " Reason: %s")
-                self.module.fail_json(msg=msg % (self.instance_id,
-                                                 lb,
-                                                 instance_state.description))
-            time.sleep(1)
-
-    def _is_instance_state_pending(self, instance_state):
-        """
-        Determines whether the instance_state is "pending", meaning there is
-        an operation under way to bring it in service.
-        """
-        # This is messy, because AWS provides no way to distinguish between
-        # an instance that is is OutOfService because it's pending vs. OutOfService
-        # because it's failing health checks. So we're forced to analyze the
-        # description, which is likely to be brittle.
-        return (instance_state and 'pending' in instance_state.description)
+            
+            if time.time() >= max_time:
+                msg = "Timeout exceeded when waiting states instances of LB {0}".format(
+                    lb.name
+                )
+                self.module.fail_json(
+                    msg=msg,
+                    wait_attempt=self.wait_attempt
+                )
+            
+            self.wait_attempt[lb.name] = self.wait_attempt[lb.name] + 1
+            # min healthcheck interval
+            time.sleep(5)
```

Comment:
this should not be hardcoded, we might want to make it a parameter that defaults to 5